pkg/image/cdi/download.go

```diff
@@ -1,19 +1,299 @@
 package cdi
 
 import (
+	"context"
+	"fmt"
+	"io"
 	"net/http"
+	"strings"
+	"time"
+
+	gocommon "github.com/harvester/go-common/common"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
 
 	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
 	"github.com/harvester/harvester/pkg/image/backend"
+	"github.com/harvester/harvester/pkg/image/common"
+	"github.com/harvester/harvester/pkg/util"
+	utilHelm "github.com/harvester/harvester/pkg/util/helm"
 )
 
 type Downloader struct {
+	httpClient http.Client
+	clientSet  *kubernetes.Clientset
+	vmio       common.VMIOperator
 }
 
-func GetDownloader() backend.Downloader {
-	return &Downloader{}
+func GetDownloader(httpClient http.Client, clientSet *kubernetes.Clientset, vmio common.VMIOperator) backend.Downloader {
+	return &Downloader{
+		httpClient: httpClient,
+		clientSet:  clientSet,
+		vmio:       vmio,
+	}
 }
 
-func (cd *Downloader) Do(_ *harvesterv1.VirtualMachineImage, _ http.ResponseWriter, _ *http.Request) error {
+func (cd *Downloader) Do(vmImg *harvesterv1.VirtualMachineImage, rw http.ResponseWriter, req *http.Request) error {
+	vmImgName := vmImg.Name
+	vmImgNamespace := vmImg.Namespace
+	podName := fmt.Sprintf("%s-%s-%s", vmImgNamespace, vmImgName, getRandomSuffix())
+
+	podTemplate := cd.generateDownlodServerPod(podName, vmImgName)
+	if _, err := cd.clientSet.CoreV1().Pods(vmImgNamespace).Create(context.TODO(), podTemplate, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("failed to create the download server pod with VM Image(%s): %w", vmImgName, err)
+	}
+	defer func() {
+		if err := cd.clientSet.CoreV1().Pods(vmImgNamespace).Delete(context.TODO(), podName, metav1.DeleteOptions{}); err != nil {
+			logrus.Errorf("Failed to delete the download server pod(%s): %v", podName, err)
+		}
+	}()
+
+	// wait pod status running
+	if err := wait.PollUntilContextTimeout(context.Background(), tickPolling, tickTimeout, true, func(context.Context) (bool, error) {
```

Comment:
we may need to make this timeout configurable. currently 60 seconds may not be enough. on a slow environment or where larger images are involved, the conversion may not complete within 60 seconds and eventually the download request times out. should we read this off a setting? 

we may also need to change how the failure is handled as in my case I get a Gateway Timeout error

Reply:
I rewrite the download mechanism.
Now, we will use the CRD to ensure the download server is ready. Then, the HTTP request will be executed after that.